streamer/src/packet.rs

```diff
@@ -81,6 +89,197 @@ pub(crate) fn recv_from(
     Ok(i)
 }
 
+/// Receive multiple messages from `sock` into buffer provided in `batch`.
+/// This is a wrapper around recvmmsg(7) call.
+#[cfg(unix)]
+pub(crate) fn recv_from(
+    batch: &mut PinnedPacketBatch,
+    socket: &UdpSocket,
+    // If max_wait is None, reads from the socket until either:
+    //   * 64 packets are read (PACKETS_PER_BATCH == 64), or
+    //   * There are no more data available to read from the socket.
+    max_wait: Option<Duration>,
+    poll_fd: &mut [PollFd],
+) -> Result<usize> {
+    // Implementation note:
+    // This is a reimplementation of the above (now, non-unix) `recv_from` function, and
+    // is explicitly meant to preserve the existing behavior, refactored for performance.
+    //
+    // This implementation is broken into two separate functions:
+    // 1. `recv_from_coalesce` - when `max_wait` is provided.
+    // 2. `recv_from_once` - when `max_wait` is not provided.
+    //
+    // This is done to avoid excessive branching in the main loop.
+
+    /// The initial socket polling timeout.
+    ///
+    /// The socket will be polled for this duration in the event that the initial
+    /// `recv_mmsg` call fails with `WouldBlock`.
+    ///
+    /// This is meant to emulate the blocking behavior of the original `recv_from` function.
+    /// The original implementation explicitly sets the socket its given as blocking, and implicitly
+    /// expects that the caller will set `socket.set_read_timeout(Some(Duration::from_millis(SOCKET_READ_TIMEOUT)))`
+    /// some time before invocation.
+    ///
+    /// Given that we are using `poll` in this implementation, and we assume the socket is set to
+    /// non-blocking, we don't need to worry about `recv_mmsg` hanging indefinitely.
+    const SOCKET_READ_TIMEOUT_MS: u16 = SOCKET_READ_TIMEOUT.as_millis() as u16;
+    use crate::streamer::SOCKET_READ_TIMEOUT;
+
+    /// Read and batch packets from the socket until batch size is [`PACKETS_PER_BATCH`] or there are no more packets to read.
+    ///
+    /// Upon calling, this will attempt to read packets from the socket, and poll for [`SOCKET_READ_TIMEOUT`]
+    /// when [`ErrorKind::WouldBlock`] is encountered.
+    ///
+    /// On subsequent iterations, when [`ErrorKind::WouldBlock`] is encountered:
+    /// - If any packets were read, the function will exit.
+    /// - If no packets were read, the function will return an error.
```

Comment:
Is this for backward compatibility? Else shoudn't we return Ok(0) if no packets are read?

Reply:
wow this makes no sense, ok 😭